sdk/src/masp.rs

```diff
@@ -1370,36 +1513,109 @@ impl<U: ShieldedUtils + MaybeSend + MaybeSync> ShieldedContext<U> {
                 return Err(Error::from(PinnedBalanceError::InvalidViewingKey));
             }
         }
-        // The address of the MASP account
-        let masp_addr = MASP;
         // Construct the key for where the transaction ID would be stored
-        let pin_key = Key::from(masp_addr.to_db_key())
-            .push(&(PIN_KEY_PREFIX.to_owned() + &owner.hash()))
-            .map_err(|_| {
-                Error::Other("Cannot obtain a storage key".to_string())
-            })?;
+        let pin_key = namada_core::types::token::masp_pin_tx_key(&owner.hash());
         // Obtain the transaction pointer at the key
         // If we don't discard the error message then a test fails,
         // however the error underlying this will go undetected
-        let txidx = rpc::query_storage_value::<C, u64>(client, &pin_key)
-            .await
-            .map_err(|_| PinnedBalanceError::NoTransactionPinned)?;
-        // Construct the key for where the pinned transaction is stored
-        let tx_key = Key::from(masp_addr.to_db_key())
-            .push(&(TX_KEY_PREFIX.to_owned() + &txidx.to_string()))
-            .map_err(|_| {
-                Error::Other("Cannot obtain a storage key".to_string())
+        let indexed_tx =
+            rpc::query_storage_value::<C, IndexedTx>(client, &pin_key)
+                .await
+                .map_err(|_| PinnedBalanceError::NoTransactionPinned)?;
+        let tx_epoch = query_epoch_at_height(client, indexed_tx.height)
+            .await?
+            .ok_or_else(|| {
+                Error::from(QueryError::General(
+                    "Queried height is greater than the last committed block \
+                     height"
+                        .to_string(),
+                ))
             })?;
-        // Obtain the pointed to transaction
-        let (tx_epoch, _tx_height, _tx_index, _tx, shielded) =
-            rpc::query_storage_value::<
-                C,
-                (Epoch, BlockHeight, TxIndex, Transfer, Transaction),
-            >(client, &tx_key)
+
+        let block = client
+            .block(indexed_tx.height.0 as u32)
             .await
-            .map_err(|_| {
-                Error::Other("Ill-formed epoch, transaction pair".to_string())
-            })?;
+            .map_err(|e| Error::from(QueryError::General(e.to_string())))?
+            .block
+            .data;
+
+        let tx = Tx::try_from(block[indexed_tx.index.0 as usize].as_ref())
```

Comment:
It seems like it might be possible that this and the code from `fetch_shielded_transfers` can be re-used here. I'm not quite sure as there is no wrapper check here, but it would be awesome if this stair-step code could be factored out into a separate function and re-used.